src/Workspaces/Core/MSBuild.BuildHost/MSBuild/ProjectFile/ProjectFile.cs

```diff
@@ -386,132 +393,75 @@ public void RemoveDocument(string filePath)
+        private MSB.Evaluation.ProjectItem FindReferenceItem(string shortAssemblyName, string fullAssemblyName, string filePath)
         {
-            if (_loadedProject is null)
-            {
-                throw new InvalidOperationException($"Unable to find reference item '{identity?.Name}'");
-            }
+            Contract.ThrowIfNull(_loadedProject, "The project was not loaded.");
 
             var references = _loadedProject.GetItems(ItemNames.Reference);
             MSB.Evaluation.ProjectItem? item = null;
 
             var fileName = Path.GetFileNameWithoutExtension(filePath);
 
-            if (identity != null)
-            {
-                var shortAssemblyName = identity.Name;
-                var fullAssemblyName = identity.GetDisplayName();
-
-                // check for short name match
-                item = references.FirstOrDefault(it => string.Compare(it.EvaluatedInclude, shortAssemblyName, StringComparison.OrdinalIgnoreCase) == 0);
+            // check for short name match
+            item = references.FirstOrDefault(it => string.Compare(it.EvaluatedInclude, shortAssemblyName, StringComparison.OrdinalIgnoreCase) == 0);
+            if (item is not null)
+                return item;
 
-                // check for full name match
-                item ??= references.FirstOrDefault(it => string.Compare(it.EvaluatedInclude, fullAssemblyName, StringComparison.OrdinalIgnoreCase) == 0);
-            }
+            // check for full name match
+            item = references.FirstOrDefault(it => string.Compare(it.EvaluatedInclude, fullAssemblyName, StringComparison.OrdinalIgnoreCase) == 0);
+            if (item is not null)
+                return item;
 
             // check for file path match
-            if (item == null)
-            {
-                var relativePath = PathUtilities.GetRelativePath(_loadedProject.DirectoryPath, filePath);
+            var relativePath = PathUtilities.GetRelativePath(_loadedProject.DirectoryPath, filePath);
 
-                item = references.FirstOrDefault(it => PathUtilities.PathsEqual(it.EvaluatedInclude, filePath)
+            item = references.FirstOrDefault(it => PathUtilities.PathsEqual(it.EvaluatedInclude, filePath)
                                                     || PathUtilities.PathsEqual(it.EvaluatedInclude, relativePath)
                                                     || PathUtilities.PathsEqual(GetHintPath(it), filePath)
                                                     || PathUtilities.PathsEqual(GetHintPath(it), relativePath));
-            }
 
-            // check for partial name match
-            if (item == null && identity != null)
+            if (item is not null)
+                return item;
+
+            var partialName = shortAssemblyName + ",";
+            var items = references.Where(it => it.EvaluatedInclude.StartsWith(partialName, StringComparison.OrdinalIgnoreCase)).ToList();
```

Comment:
>Where [](http://example.com/codeflow?start=35&length=5)

another nit, but changing this to just use Single instead of Where/ToList/Count/throw would be easier to read

Reply:
Going to defer this to a follow-up PR, if for no other reason then there aren't many existing tests in this area, so I want to make sure we've got the cases well tested first before we tinker.